File: battle/data/CustomEntity.java
package common.battle.data;

import common.io.json.JsonClass;
import common.io.json.JsonClass.NoTag;
import common.io.json.JsonDecoder;
import common.io.json.JsonField;
import common.io.json.JsonField.GenType;
import common.util.Data;
import common.util.unit.Trait;

import java.util.ArrayList;
import java.util.List;

@JsonClass(noTag = NoTag.LOAD)
public abstract class CustomEntity extends DataEntity {

	@JsonField(gen = GenType.GEN)
	public AtkDataModel rep, rev, res, cntr;

	@JsonField(gen = GenType.GEN, usePool = true)
	public AtkDataModel[] atks;

	public int tba, base, touch = TCH_N;
	public boolean common = true;

	/**
	 * This field is used to filter all the procs of units if common is false,
	 * Also used for counter
	 */
	@JsonField(block = true)
	private Proc all;

	@Override
	public int allAtk() {
		int ans = 0, temp = 0, c = 1;
		for (AtkDataModel adm : atks)
			if (adm.pre > 0) {
				ans += temp / c;
				temp = adm.getDire() > 0 ? adm.atk : 0;
				c = 1;
			} else {
				temp += adm.getDire() > 0 ? adm.atk : 0;
				c++;
			}
		ans += temp / c;
		return ans;
	}

	/**
	 * Updates the procs in all and initializes if it is null
	 */
	public void updateAllProc() {
		all = Proc.blank();
		for (int i = 0; i < Data.PROC_TOT; i++) {
			if (Data.procSharable[i]) {
				all.getArr(i).set(getProc().getArr(i));
			} else
				for (AtkDataModel adm : atks)
					if (!all.getArr(i).exists())
						all.getArr(i).set(adm.proc.getArr(i));
		}
	}

	/**
	 * Gets all procs for units without common proc
	 */
	@Override
	public Proc getAllProc() {
		if (common)
			return getProc();
		if (all == null)
			updateAllProc();
		return all;
	}

	@Override
	public int getAtkCount() {
		return atks.length;
	}

	@Override
	public MaskAtk getAtkModel(int ind) {
		if (ind < atks.length)
			return atks[ind];
		if (ind == atks.length)
			return rev;
		if (ind == atks.length + 1)
			return res;
		return null;
	}

	@Override
	public MaskAtk[] getAtks() {
		return atks;
	}

	public String getAvailable(String str) {
		while (contains(str))
			str += "'";
		return str;
	}

	@Override
	public int getItv() {
		int longPre = 0;
		for (AtkDataModel adm : atks)
			longPre += adm.pre;
		return longPre + Math.max(getTBA() - 1, getPost());
	}

	@Override
	public int getPost() {
		int ans = getAnimLen();
		for (AtkDataModel adm : atks)
			ans -= adm.pre;
		return ans;
	}

	@Override
	public Proc getProc() {
		return rep.getProc();
	}

	@Override
	public MaskAtk getRepAtk() {
		return rep;
	}

	@Override
	public AtkDataModel getResurrection() {
		return res;
	}

	@Override
	public AtkDataModel getRevenge() {
		return rev;
	}

	@Override
	public AtkDataModel getCounter() { return cntr; }

	@Override
	public int getTBA() {
		return tba;
	}

	@Override
	public int getTouch() {
		return touch;
	}

	public void importData(MaskEntity de) {
		hp = de.getHp();
		hb = de.getHb();
		speed = de.getSpeed();
		range = de.getRange();
		abi = de.getAbi();
		loop = de.getAtkLoop();
		traits = new ArrayList<>();
		for(Trait t : de.getTraits()) {
			if(!t.BCTrait)
				traits.add(t);
			else if(t.id.id != Data.TRAIT_EVA && t.id.id != Data.TRAIT_WITCH)
				traits.add(t);
		}
		width = de.getWidth();
		tba = de.getTBA();
		touch = de.getTouch();
		death = de.getDeathAnim();
		will = de.getWill();
		if (de instanceof CustomEntity) {
			importData$1((CustomEntity) de);
			return;
		}

		base = de.touchBase();
		common = ((DefaultData)de).isCommon();
		rep = new AtkDataModel(this);
		rep.proc = de.getRepAtk().getProc().clone();
		int m = de.getAtkCount();
		atks = new AtkDataModel[m];
		for (int i = 0; i < m; i++) {
			atks[i] = new AtkDataModel(this, de, i);
			for (int j : BCShareable)
				atks[i].proc.getArr(j).set(de.getProc().getArr(j));
		}
	}

	@Override
	public boolean isLD() {
		boolean ans = false;
		for (AtkDataModel adm : atks)
			ans |= adm.isLD();
		if(getRevenge() != null)
			ans |= getRevenge().isLD();
		if(getResurrection() != null)
			ans |= getResurrection().isLD();
		return ans;
	}

	/**
	 * Returns if a specific attack is LD,
	 * used to handle LD units that have an attack that isn't LD properly
	 * @param ind The attack to get.
	 */
	@Override
	public boolean isLD(int ind) {
		if (ind == atks.length)
			return rev.isLD();
		if (ind == atks.length + 1)
			return res.isLD();
		return atks[ind].isLD();
	}

	@Override
	public boolean isOmni() {
		boolean ans = false;
		for (AtkDataModel adm : atks)
			ans |= adm.isOmni();
		if(getRevenge() != null)
			ans |= getRevenge().isOmni();
		if(getResurrection() != null)
			ans |= getResurrection().isOmni();
		return ans;
	}

	/**
	 * Returns if a specific attack is Omni,
	 * used to handle Omni units that have an attack that lacks omni properly
	 * @param ind The attack to get.
	 */
	@Override
	public boolean isOmni(int ind) {
		if (ind == atks.length)
			return rev.isOmni();
		if (ind == atks.length + 1)
			return res.isOmni();
		return atks[ind].isOmni();
	}

	@Override
	public boolean isRange() {
		for (AtkDataModel adm : atks)
			if (adm.range)
				return true;
		return false;
	}

	@Override
	public int[][] rawAtkData() {
		int[][] ans = new int[atks.length][];
		for (int i = 0; i < atks.length; i++)
			ans[i] = atks[i].getAtkData();
		return ans;
	}

	@Override
	public int touchBase() {
		return base == 0 ? range : base;
	}

	private boolean contains(String str) {
		if (atks == null || atks.length == 0)
			return false;
		for (AtkDataModel adm : atks)
			if (adm != null && adm.str.equals(str))
				return true;
		return false;
	}

	private void importData$1(CustomEntity ce) {
		base = ce.base;
		common = ce.common;
		rep = new AtkDataModel(this, ce.rep);

		List<AtkDataModel> temp = new ArrayList<>();
		List<AtkDataModel> tnew = new ArrayList<>();
		int[] inds = new int[ce.atks.length];
		for (int i = 0; i < ce.atks.length; i++) {
			if (!temp.contains(ce.atks[i])) {
				temp.add(ce.atks[i]);
				tnew.add(new AtkDataModel(this, ce.atks[i]));
			}
			inds[i] = temp.indexOf(ce.atks[i]);
		}
		atks = new AtkDataModel[ce.atks.length];
		for (int i = 0; i < atks.length; i++) {
			atks[i] = tnew.get(inds[i]);

			if (!ce.getClass().equals(getClass()))
				atks[i].getProc().SUMMON.id = null;
		}
	}

	@JsonDecoder.OnInjected
	public void onInjected() {
		for (int i = 0; i < traits.size(); i++)
			if (traits.get(i) == null) {
				traits.remove(i);
				i--;
			}
	}
}

File: battle/data/CustomEntity.kt
package common.battle.data

import common.io.InStream
import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonClass
import common.io.json.JsonClass.NoTag
import common.io.json.JsonEncoder
import common.io.json.JsonField
import common.io.json.JsonField.GenType
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.Data.Proc
import common.util.pack.Soul
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import java.util.*

@JsonClass(noTag = NoTag.LOAD)
abstract class CustomEntity : DataEntity() {
    @JsonField(gen = GenType.GEN)
    var rep: AtkDataModel? = null

    @JsonField(gen = GenType.GEN)
    var rev: AtkDataModel? = null

    @JsonField(gen = GenType.GEN)
    var res: AtkDataModel? = null

    @JsonField(gen = GenType.GEN, usePool = true)
    var atks: Array<AtkDataModel?>?
    var tba = 0
    var base = 0
    var touch: Int = common.util.Data.Companion.TCH_N
    var common = true
    private var all: Proc? = null
    override fun allAtk(): Int {
        var ans = 0
        var temp = 0
        var c = 1
        for (adm in atks!!) if (adm!!.pre > 0) {
            ans += temp / c
            temp = if (adm!!.getDire() > 0) adm!!.atk else 0
            c = 1
        } else {
            temp += if (adm!!.getDire() > 0) adm!!.atk else 0
            c++
        }
        ans += temp / c
        return ans
    }

    override fun getAllProc(): Proc? {
        if (all != null) return all
        all = rep!!.getProc()!!.clone()
        for (adm in atks!!) {
            for (i in 0 until common.util.Data.Companion.PROC_TOT) if (!all.getArr(i).exists()) all.getArr(i).set(adm!!.proc!!.getArr(i))
        }
        return all
    }

    fun getAllProcs(): Array<Proc?> {
        val n = atks!!.size + 1
        val ans: Array<Proc?> = arrayOfNulls<Proc>(n)
        ans[0] = rep!!.proc
        for (i in atks!!.indices) ans[i + 1] = atks!![i]!!.proc
        return ans
    }

    override fun getAtkCount(): Int {
        return atks!!.size
    }

    override fun getAtkModel(ind: Int): MaskAtk? {
        if (ind < atks!!.size) return atks!![ind]
        if (ind == atks!!.size) return rev
        return if (ind == atks!!.size + 1) res else null
    }

    fun getAvailable(str: String): String {
        var str = str
        while (contains(str)) str += "'"
        return str
    }

    override fun getItv(): Int {
        return getAnimLen() + tba
    }

    override fun getPost(): Int {
        var ans: Int = getAnimLen()
        for (adm in atks!!) ans -= adm!!.pre
        return ans
    }

    override fun getProc(): Proc? {
        return rep!!.getProc()
    }

    override fun getRepAtk(): MaskAtk? {
        return rep
    }

    override fun getResurrection(): AtkDataModel? {
        return res
    }

    override fun getRevenge(): AtkDataModel? {
        return rev
    }

    override fun getTBA(): Int {
        return tba
    }

    override fun getTouch(): Int {
        return touch
    }

    open fun importData(de: MaskEntity) {
        hp = de.getHp()
        hb = de.getHb()
        speed = de.getSpeed()
        range = de.getRange()
        abi = de.getAbi()
        loop = de.getAtkLoop()
        type = de.getType()
        width = de.getWidth()
        shield = de.getShield()
        tba = de.getTBA()
        touch = de.getTouch()
        death = de.getDeathAnim()
        if (de is CustomEntity) {
            `importData$1`(de)
            return
        }
        base = de.touchBase()
        common = false
        rep = AtkDataModel(this)
        rep.proc = de.getRepAtk()!!.getProc().clone()
        val m: Int = de.getAtkCount()
        atks = arrayOfNulls<AtkDataModel>(m)
        for (i in 0 until m) atks!![i] = AtkDataModel(this, de, i)
    }

    override fun isLD(): Boolean {
        var ans = false
        for (adm in atks!!) ans = ans or adm!!.isLD()
        return ans
    }

    override fun isOmni(): Boolean {
        var ans = false
        for (adm in atks!!) ans = ans or adm!!.isOmni()
        return ans
    }

    override fun isRange(): Boolean {
        for (adm in atks!!) if (adm!!.range) return true
        return false
    }

    override fun rawAtkData(): Array<IntArray?> {
        val ans = arrayOfNulls<IntArray>(atks!!.size)
        for (i in atks!!.indices) ans[i] = atks!![i]!!.getAtkData()
        return ans
    }

    override fun touchBase(): Int {
        return if (base == 0) range else base
    }

    protected fun zreada(`is`: InStream) {
        val ver: Int = common.util.Data.Companion.getVer(`is`.nextString())
        if (ver >= 404) `zreada$000404`(`is`)
    }

    private operator fun contains(str: String): Boolean {
        if (atks == null || atks!!.size == 0) return false
        for (adm in atks!!) if (adm != null && adm.str == str) return true
        return false
    }

    private fun `importData$1`(ce: CustomEntity) {
        base = ce.base
        common = ce.common
        rep = AtkDataModel(this, ce.rep)
        val temp: MutableList<AtkDataModel?> = ArrayList<AtkDataModel?>()
        val tnew: MutableList<AtkDataModel> = ArrayList<AtkDataModel>()
        val inds = IntArray(ce.atks!!.size)
        for (i in ce.atks!!.indices) {
            if (!temp.contains(ce.atks!![i])) {
                temp.add(ce.atks!![i])
                tnew.add(AtkDataModel(this, ce.atks!![i]))
            }
            inds[i] = temp.indexOf(ce.atks!![i])
        }
        atks = arrayOfNulls<AtkDataModel>(ce.atks!!.size)
        for (i in atks!!.indices) atks!![i] = tnew[inds[i]]
    }

    private fun `zreada$000404`(`is`: InStream) {
        hp = `is`.nextInt()
        hb = `is`.nextInt()
        speed = `is`.nextInt()
        range = `is`.nextInt()
        abi = `is`.nextInt()
        type = `is`.nextInt()
        width = `is`.nextInt()
        shield = `is`.nextInt()
        tba = `is`.nextInt()
        base = `is`.nextInt()
        touch = `is`.nextInt()
        loop = `is`.nextInt()
        death = common.pack.PackData.Identifier.Companion.parseInt<Soul>(`is`.nextInt(), Soul::class.java)
        common = `is`.nextInt() > 0
        rep = AtkDataModel(this, `is`)
        val m: Int = `is`.nextInt()
        val set: Array<AtkDataModel?> = arrayOfNulls<AtkDataModel>(m)
        for (i in 0 until m) set[i] = AtkDataModel(this, `is`)
        val n: Int = `is`.nextInt()
        atks = arrayOfNulls<AtkDataModel>(n)
        for (i in 0 until n) atks!![i] = set[`is`.nextInt()]
        val adi: Int = `is`.nextInt()
        if (adi and 1 > 0) rev = AtkDataModel(this, `is`)
        if (adi and 2 > 0) res = AtkDataModel(this, `is`)
    }
}
